tools/dataset.cpp

```diff
@@ -1 +1,1305 @@
+}
+
+void export_csv_d( //
+    ukv_docs_export_t& c,
+    std::vector<ptr_range_gt<ukv_key_t const>> const& keys,
+    fields_t const& strided_fields,
+    ukv_size_t size_in_bytes,
+    vals_t const& values) {
+
+    ukv_size_t size = 0;
+    arrow::Status status;
+
+    for (auto _ : keys)
+        size += _.size();
+
+    arrow::NumericBuilder<arrow::Int64Type> int_builder;
+    arrow::StringBuilder str_builder;
+
+    status = int_builder.Resize(size);
+    return_if_error(status.ok(), c.error, 0, "Can't instatinate builder");
+    status = str_builder.Resize(size);
+    return_if_error(status.ok(), c.error, 0, "Can't instatinate builder");
+
+    std::shared_ptr<arrow::Array> keys_array;
+    std::shared_ptr<arrow::Array> docs_array;
+    std::vector<ukv_key_t> keys_vec;
+    std::vector<std::string> docs_vec;
+
+    keys_vec.reserve(size);
+    docs_vec.reserve(size);
+
+    if (c.fields) {
+        std::vector<std::string> fields(c.fields_count);
+        for (size_t idx = 0; idx < c.fields_count; ++idx)
+            fields[idx] = strided_fields[idx];
+
+        sub_content(values, keys, fields, &docs_vec, &keys_vec, nullptr, nullptr, 1);
+    }
+    else
+        whole_content(values, keys, &docs_vec, &keys_vec, nullptr, nullptr, 1);
+
+    status = int_builder.AppendValues(keys_vec);
+    return_if_error(status.ok(), c.error, 0, "Can't append keys");
+    status = int_builder.Finish(&keys_array);
+    return_if_error(status.ok(), c.error, 0, "Can't finish array(keys)");
+
+    status = str_builder.AppendValues(docs_vec);
+    return_if_error(status.ok(), c.error, 0, "Can't append docs");
+    status = str_builder.Finish(&docs_array);
+    return_if_error(status.ok(), c.error, 0, "Can't finish array(docs)");
+
+    arrow::FieldVector fields;
+    fields.push_back(arrow::field("id", arrow::int64()));
+    fields.push_back(arrow::field("doc", arrow::int64()));
+    std::shared_ptr<arrow::Schema> schema = std::make_shared<arrow::Schema>(fields);
+
+    std::shared_ptr<arrow::Table> table;
+    table = arrow::Table::Make(schema, {keys_array, docs_array});
+
+    char file_name[uuid_length];
+    make_uuid(file_name);
+
+    auto maybe_outstream = arrow::io::FileOutputStream::Open(fmt::format("{}{}", file_name, c.paths_extension));
+    return_if_error(maybe_outstream.ok(), c.error, 0, "Can't open file");
+    std::shared_ptr<arrow::io::FileOutputStream> outstream = *maybe_outstream;
+
+    status = arrow::csv::WriteCSV(*table, arrow::csv::WriteOptions::Defaults(), outstream.get());
+    return_if_error(status.ok(), c.error, 0, "Can't write in file");
+}
+
+void export_ndjson_d( //
+    ukv_docs_export_t& c,
+    std::vector<ptr_range_gt<ukv_key_t const>> const& keys,
+    fields_t const& strided_fields,
+    ukv_size_t size_in_bytes,
+    vals_t const& values) {
+
+    char file_name[uuid_length];
+    make_uuid(file_name);
+    std::ofstream output(fmt::format("{}{}", file_name, c.paths_extension));
```

Comment:
Please, never use `ofstream` or the rest of the old-school IO libraries from STL. Especially on the hot path.